src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Cohost/CohostOnTypeFormattingEndpoint.cs

```diff
@@ -0,0 +1,176 @@
+            return null;
+        }
+
+        // We have to go to OOP to find out if we want Html formatting for this request. This is a little unfortunate
+        // but just asking Html for formatting, just in case, would be bad for a couple of reasons. Firstly, the Html
+        // trigger characters are a superset of the C# triggers, so we can't use that as a sign. Secondly, whilst we
+        // might be making one Html request, it could be then calling CSS or TypeScript servers, so our single request
+        // to OOP could potentially save a few requests downstream. Lastly, our request to OOP is MessagePack which is
+        // generally faster than Json anyway.
+        var triggerKind = await _remoteServiceInvoker.TryInvokeAsync<IRemoteFormattingService, IRemoteFormattingService.TriggerKind>(
+            razorDocument.Project.Solution,
+            (service, solutionInfo, cancellationToken) => service.GetOnTypeFormattingTriggerKindAsync(solutionInfo, razorDocument.Id, request.Position.ToLinePosition(), request.Character, cancellationToken),
+            cancellationToken).ConfigureAwait(false);
+
+        if (triggerKind == IRemoteFormattingService.TriggerKind.Invalid)
+        {
+            return null;
+        }
+
+        var sourceText = await razorDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+        ImmutableArray<TextChange> htmlChanges = [];
+        if (triggerKind == IRemoteFormattingService.TriggerKind.ValidHtml)
+        {
+            _logger.LogDebug($"Getting Html formatting changes for {razorDocument.FilePath}");
+            var htmlResult = await GetHtmlFormattingEditsAsync(request, razorDocument, cancellationToken).ConfigureAwait(false);
+
+            if (htmlResult is not { } htmlEdits)
+            {
+                // We prefer to return null, so the client will try again
+                _logger.LogDebug($"Didn't get any edits back from Html");
+                return null;
+            }
+
+            htmlChanges = htmlEdits.SelectAsArray(sourceText.GetTextChange);
+        }
+
+        var options = RazorFormattingOptions.From(request.Options, clientSettings.AdvancedSettings.CodeBlockBraceOnNextLine);
+
+        _logger.LogDebug($"Calling OOP with the {htmlChanges.Length} html edits, so it can fill in the rest");
+        var remoteResult = await _remoteServiceInvoker.TryInvokeAsync<IRemoteFormattingService, ImmutableArray<TextChange>>(
+            razorDocument.Project.Solution,
+            (service, solutionInfo, cancellationToken) => service.GetOnTypeFormattingEditsAsync(solutionInfo, razorDocument.Id, htmlChanges, request.Position.ToLinePosition(), request.Character, options, cancellationToken),
+            cancellationToken).ConfigureAwait(false);
+
+        if (remoteResult is [_, ..] allChanges)
+        {
+            _logger.LogDebug($"Got a total of {allChanges.Length} ranges back from OOP");
+
+            return allChanges.Select(sourceText.GetTextEdit).ToArray();
```

Comment:
It feels like we need an extension method for a `Select(...)` followed by `ToArray()` on an `ImmutableArray<T>`. Essentially, it should be safe to do something like the following, and we should add an extension method for that pattern to avoid an extra copy.

```C#
var tempArray = allChanges.SelectAsArray(sourceText.GetTextEdit);
return ImmutableCollectionsMarshal.AsArray(tempArray);
```

Reply:
Sure, but what do we call it when `SelectAsArray` is already taken 😛

Also even doing that seems like a roundabout way to get at an array (ie, use a builder to create an immutable array to get at its inner array). Perhaps just a `ConvertToArray` that allocates an array and fills it in would be just as good?

Reply:
Yeah, I've been thinking about that as well. Note, I wasn't asking for any change here. Just wanted to make a note.

Reply:
I've ran into a couple more places where this would useful, in converting to TextChange. I've made a note for myself to check if those survive the rest of the work too. Naming is seriously hard though :)

Reply:
Actually, I don't think this one is too hard because there's at least a name precedent with `Array.ConvertAll`. However, we should come up with a good naming pattern than makes sense and is inherently readable. `SelectAsArray` was originally picked because that's the naming scheme used over in Roslyn. However, it's not ideal in the LSP world where regular ol' arrays are the norm.

Reply:
`ConvertToArray` is definitely leading contender for naming right now :)

Reply:
I'm not a 100% sure it represents the right pattern though. Is "ToArray" the right suffix? Why was "AsArray" picked for the others? "AsArray" seems inconsistent with how it's used by APIs, such as `string.AsSpan()`, `ImmutableCollectionsMarshal.AsArray(...)`, `ImmutableArray<T>.AsSpan(...)` , etc. It seems like "As" is used to mean "treat the exact same memory as something else". However, "To" definitely implies making the subject into another thing.

I need to think on this for awhile.

Reply:
By that logic, isn't `SelectAsArray` also wrong, since it converts? Maybe it should just `SelectAsMutableArray`, that way at least immutable collections will still be the default.

Reply:
> By that logic, isn't `SelectAsArray` also wrong

Yup! That's what I meant to imply. It was only picked because that's what some of the internal helpers were called in Roslyn, but I'm not 100% sure that's best.

Reply:
I always rationalized it as talking about the result of the Select, not the things being Selected. So its, "Run Select(), and then give me the result as an immutable array". The first half is the thing that changes the data, and the As applies to the last half. Thats definitely reverse engineered justification though.

Reply:
Oh, I definitely get that. I'm just not sure that I want to propagate that if `ConvertToArray` is the front running. After all, `Convert` is just `Select`.